torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -513,12 +513,27 @@ class CppMicroGemmAMX(CppMicroGemm):
     {{kernel.assert_function}}(K % 2 == 0, "K dimension must be multiple of 2");
 {%- if use_cached_dequantized_B %}
     // Create a stack-allocated buffer for tiles of B.
-    // Except maybe for the tail-case, an AMX tile of B has 16x32 BF16 elements,
-    // but to make things simple, we'll allocate space for a 16x32 tile even for the tail.
-    // It's possible that N may not be equal to Nr, but might be a multiple of it.
-    // So, we'll allocate the buffer accordingly.
+    // Except maybe for the tail-case, an AMX tile of B has 16x32 BF16 elements.
     const auto num_elements_per_b_tile = 512;
-    const auto buf_size_per_nr_block = ((K + {{block_k}} - 1) / {{block_k}}) * num_elements_per_b_tile * 2;
+    const auto last_k_offset = K / {{block_k}} * {{block_k}};
+    const auto tail_k_size = K - last_k_offset;
+    const auto buf_size_per_nr_block = ((K / {{block_k}}) * num_elements_per_b_tile + tail_k_size * 32) * 2;
+    // Although the outer loop in the micro-kernel iterates on n, we could only cache K * Nr elements of
+    // dequantized B, pertaining to each value of n.
+    // But we allow for caching of multiple K * Nr sets in the code to support N != Nr for two reasons:
+    // 1. If N would not be equal to Nr in the future, outer loop in the micro-kernel that iterates over n
+    //    could potentially be unrolled. If the cache could only store K * Nr elements, loop-unrolling would be
+    //    impossible due to memory aliasing.
```

Comment:
This note explains why the cache is sized `N x K` instead of `Nr x K`.

 Since `n` is in the outer-loop, unrolling it would entail multiple iterations of the outer loop, with each iteration using _different weights_. However, if the dequantized weights buffer would be of size `Nr x K` (as opposed to `N x K`, which is essentially `some_natural_number x Nr x K`), then loop-unrolling of the outer loop won't be possible due to aliasing of the `dequantized B` buffer.